### PR TITLE
Add support for publishing a single program.

### PR DESCRIPTION
**Pull request by reemaka**
### Description

Allow publishing a single program at a time. Also publishes any of the program's edited questions. If any of the modified questions are shared with other programs, the program is ineligible for being published individually.

UI is tentative; just added this to the dropdown on the program index page for now. I also plan to add a confirmation dialog in a follow-up PR. Right now this just makes the change as soon as the admin hits publish.

## Release notes

The title of the pull request will be used as the default release notes description. If more detail is needed to communicate to partners the scope of the PR's changes, use this release notes section.

#### General

- [x] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release >

#### New Features

- [x] Add new FeatureFlag env vars to `server/conf/application.conf`
- [x] Conditioned new functionality on a [FeatureFlag](https://docs.civiform.us/contributor-guide/developer-guide/feature-flags)

### Issue(s) this completes

Part of #1589

### CODE REVIEW

File: server/app/views/admin/programs/ProgramIndexView.java [hunk 1]
```diff
@@ -328,6 +328,9 @@ private ProgramCardFactory.ProgramCardData buildProgramCardData(
         draftRowExtraActions.add(maybeManageTranslationsLink.get());
       }
       draftRowExtraActions.add(renderEditStatusesLink(draftProgram.get()));
+      if (featureFlags.getFlagEnabled(request, PUBLISH_SINGLE_PROGRAM_ENABLED)) {
+        draftRowExtraActions.add(renderPublishProgramLink(draftProgram.get(), request));
```
Comment (bion) [inline server/app/views/admin/programs/ProgramIndexView.java:332]: **suggestion (non-blocking):** (you mention that the UI is tentative and this is feature flagged so this is non-blocking) could disable the button if the program has modified questions shared with other programs

File: server/conf/helper/feature-flags.conf [hunk 2]
```diff
@@ -48,3 +48,5 @@ phone_question_type_enabled = ${?PHONE_QUESTION_TYPE_ENABLED}
 bypass_login_language_screens = false
 bypass_login_language_screens = ${?BYPASS_LOGIN_LANGUAGE_SCREENS}
 
+publish_single_program_enabled = false
+publish_single_program_enabled = ${?PUBLISH_SINGLE_PROGRAM_ENABLED}
```
Comment (bion) [inline server/conf/helper/feature-flags.conf:52]: **nitpick:** missing newline at end of file

**Review by bion — CHANGES_REQUESTED** [review]

File: server/app/repository/VersionRepository.java [hunk 3]
```diff
@@ -165,6 +173,119 @@ private Version publishNewSynchronizedVersion(PublishMode publishMode) {
...
+      existingDraft.refresh();
+      active.refresh();
+      newDraft.refresh();
+      transaction.commit();
+    } catch (NonUniqueResultException | SerializableConflictException | RollbackException e) {
+      transaction.rollback(e);
+      transaction.end();
+      publishNewSynchronizedVersion(programToPublishAdminName);
+    } finally {
+      transaction.end();
```
[the remark this answers is not in the document]
Comment (reemaka) [reply server/app/repository/VersionRepository.java:285]: Nope, added a comment to explain (well, copied a comment from getDraftVersion)

File: server/app/repository/VersionRepository.java [hunk 4]
```diff
@@ -165,6 +173,119 @@ private Version publishNewSynchronizedVersion(PublishMode publishMode) {
...
+      Version existingDraft = getDraftVersion();
+      Version active = getActiveVersion();
+      // Any drafts not being published right now will be moved to newDraft.
+      Version newDraft = new Version(LifecycleStage.DRAFT);
+      database.insert(newDraft);
+
+      Optional<Program> programToPublish =
+          existingDraft.getProgramByName(programToPublishAdminName);
+      Preconditions.checkState(
+          programToPublish.isPresent(), "Program being published not found in Draft.");
```
[the remark this answers is not in the document]
Comment (reemaka) [reply server/app/repository/VersionRepository.java]: Done

File: server/app/repository/VersionRepository.java [hunk 5]
```diff
@@ -165,6 +173,119 @@ private Version publishNewSynchronizedVersion(PublishMode publishMode) {
...
+                  ImmutableMap.toImmutableMap(
+                      QuestionDefinition::getId, QuestionDefinition::getName));
+      ImmutableSet<String> questionsToPublishNames =
+          programToPublish.get().getProgramDefinition().blockDefinitions().stream()
+              .map(BlockDefinition::programQuestionDefinitions)
+              .flatMap(ImmutableList::stream)
+              .map(ProgramQuestionDefinition::id)
+              .filter(questionIdToNameLookup::containsKey)
+              .map(questionIdToNameLookup::get)
+              .collect(ImmutableSet.toImmutableSet());
```
[the remark this answers is not in the document]
Comment (reemaka) [reply server/app/repository/VersionRepository.java]: Done

File: server/app/repository/VersionRepository.java [hunk 6]
```diff
@@ -165,6 +173,119 @@ private Version publishNewSynchronizedVersion(PublishMode publishMode) {
...
+      // We only need to look at draft programs because if a question has been modified, any
+      // programs that reference it will have been added to the draft.
+      ImmutableMap<String, ImmutableSet<ProgramDefinition>> referencingDraftProgramsByQuestionName =
+          buildReferencingProgramsMap(existingDraft);
+      boolean anyDraftQuestionIsShared =
+          questionsToPublishNames.stream()
+              .anyMatch(
+                  questionName ->
+                      referencingDraftProgramsByQuestionName.containsKey(questionName)
+                          && referencingDraftProgramsByQuestionName.get(questionName).size() > 1);
```
[the remark this answers is not in the document]
Comment (reemaka) [reply server/app/repository/VersionRepository.java]: Done

File: server/app/repository/VersionRepository.java [hunk 7]
```diff
@@ -165,6 +173,119 @@ private Version publishNewSynchronizedVersion(PublishMode publishMode) {
...
+      // Move forward the ACTIVE version.
+      active.setLifecycleStage(LifecycleStage.OBSOLETE);
+      existingDraft.setLifecycleStage(LifecycleStage.ACTIVE);
+
+      existingDraft.save();
+      active.save();
+      newDraft.save();
+      existingDraft.refresh();
+      active.refresh();
+      newDraft.refresh();
```
[the remark this answers is not in the document]
Comment (reemaka) [reply server/app/repository/VersionRepository.java]: Done

File: server/app/repository/VersionRepository.java [hunk 8]
```diff
@@ -165,6 +173,119 @@ private Version publishNewSynchronizedVersion(PublishMode publishMode) {
...
+      transaction.commit();
+    } catch (NonUniqueResultException | SerializableConflictException | RollbackException e) {
+      transaction.rollback(e);
+      transaction.end();
+      publishNewSynchronizedVersion(programToPublishAdminName);
```
[the remark this answers is not in the document]
Comment (reemaka) [reply server/app/repository/VersionRepository.java:269]: Done

File: server/app/services/program/CantPublishProgramWithSharedQuestionsException.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package services.program;
+
+public class CantPublishProgramWithSharedQuestionsException extends Exception {
```
[the remark this answers is not in the document]
Comment (reemaka) [reply server/app/services/program/CantPublishProgramWithSharedQuestionsException.java:8]: Done

File: server/app/repository/VersionRepository.java [hunk 10]
```diff
@@ -165,6 +173,119 @@ private Version publishNewSynchronizedVersion(PublishMode publishMode) {
     }
   }
 
+  /**
+   * Publish the specified DRAFT program and its modified questions. No other programs/questions
+   * will be published. The DRAFT program and its DRAFT questions will become ACTIVE. The ACTIVE
+   * version of all other programs and questions will be copied to the new ACTIVE version. The DRAFT
+   * version of all other programs and questions will be copied to a new DRAFT version.
+   *
+   * @throws CantPublishProgramWithSharedQuestionsException if any of the program's modified
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/repository/VersionRepository.java:183]: oh, thanks!

[the remark this answers is not in the document]
Comment (bion) [reply server/app/repository/VersionRepository.java:183]: thanks!

**Review by bion — APPROVED** [review]

**Pull request merged**